### ISSUE TITLE
Can't detect bomb damage

### ISSUE THREAD

**Issue by xavier-rodet** [markus-wa/demoinfocs-golang#169]
**Describe the bug**
`PlayerHurt.Weapon` is unkown when a player is receiving bomb explosion damage

**To Reproduce**
At the end of round 25 of this demo, 2 players are receiving bomb damage : https://gofile.io/?c=Je2o0z

Code:
```go
parser.RegisterEventHandler(func(e events.MatchStart) {
    round := 1

    parser.RegisterEventHandler(func(e events.RoundStart) {
        round++
    })

    parser.RegisterEventHandler(func(e events.PlayerHurt) {
        if 25 == round {
            fmt.Printf("[PlayerHurt] %s receive %d damage from weapon %s\n", e.Player.Name, e.HealthDamage, e.Weapon.String())
        }
    })
})
```

Current result: 
```
demo-analyzer-api    | [PlayerHurt] [B]4ll receive 277 damage from weapon UNKNOWN
demo-analyzer-api    | [PlayerHurt] Luckso receive 7 damage from weapon UNKNOWN
demo-analyzer-api exited with code 0
```

**Expected behavior**
The weapon should be the bomb

**Comment by BestAwperEver** [comment]
I suppose @markus-wa will correct me if I'm wrong, but AFAIK you cannot tell if a player was damaged by a bomb explosion from the event itself.

Here is a reference to the raw data from which `PlayerHurt` structure gets constructed:
https://wiki.alliedmods.net/Counter-Strike:_Global_Offensive_Events#player_hurt

If a player got damage from the bomb explosion, it is referenced as "world damage" and therefore the weapon id is 0, which corresponds to `EqUnknown == 0` in library terms.

You should be able (though not with 100% accuracy) to tell if a player got bomb explosion damage by tracking `BombExploded` event and store a `bool` value somewhere inside your application which would indicate that the bomb has exploded in the current tick. Then you can check the value inside `PlayerHurt` event if you get unknown weapon.

When I say that this is not 100% accurate I mean that you can get some (although quite improbable) collisions when, for example, a player gets world damage from fall in the exact same tick as bomb has exploded. Or may be your demo will be corrupted and the event of bomb explosion will be absent (from my expierience that happens from time to time even in official GOTV matchmaking demos, and by "that" I mean the situation when a demo is missing some game data which should be recorded). But in most cases you will be just fine.

Still this inconsistency doesn't allow this method to be built in the library core, that's why (imo) it is not implemented.

**Comment by xavier-rodet** [comment]
Thanks for these informations.

I was thinking about doing exactly the way you're explaining, but I was also wondering if it could be directly implemented in the library, however from what you said I guess it's indeed a bad idea as it will alter the way a demo really "behave".

I guess that if @markus-wa agreed with this conclusion, he could close this issue!

**Comment by markus-wa** [comment]
@BestAwperEver is right that we can't twll for sure. However we could implement functionality that all `unknown` damage taken at the bomb explosion tick is counted as bomb damage.

Usually I'd be against this type of hack but the only possible false positive I can see is when a player takes fall damage during exactly that tick. Considering that that's a pretty small downside imo I'd be open to include this functionality. Possibly even with a bomb radius check, but that would be secondary.

But I might give it some more thought first

**Comment by BestAwperEver** [comment]
Note: in some demos `BombExploded` event  _consistently_ gets dispatched after players got hurt from the explosion (e.g. [this](https://www.hltv.org/download/demo/18435), overpass) and in some other it is dispatched in the right order -- first explosion, then `PlayerHurt` (see [this demo](https://www.hltv.org/download/demo/54175), dust2).

Sample output (overpass):
```
DEBUG: tick 184107: Bomb has been planted on A by pronax * Kinguin
DEBUG: tick 185395: (nil) hurt OLOFMEISTER * Kinguin with UNKNOWN in Head (92 hp left, 8 damage)
DEBUG: tick 188715: (nil) hurt pronax * Kinguin with UNKNOWN in Generic (96 hp left, 4 damage)
DEBUG: tick 188715: pronax * Kinguin got 4 world (probably fall) damage
DEBUG: tick 188715: (nil) hurt seang@res with UNKNOWN in Generic (92 hp left, 8 damage)
DEBUG: tick 188715: seang@res got 8 world (probably fall) damage
DEBUG: tick 188715: (nil) hurt n0thing^G with UNKNOWN in Generic (76 hp left, 8 damage)
DEBUG: tick 188715: n0thing^G got 8 world (probably fall) damage
DEBUG: tick 188715: (nil) hurt Skadoodle with UNKNOWN in Generic (21 hp left, 27 damage)
DEBUG: tick 188715: Skadoodle got 27 world (probably fall) damage
DEBUG: tick 188715: (nil) hurt flushamad * Kinguin with UNKNOWN in Generic (92 hp left, 8 damage)
DEBUG: tick 188715: flushamad * Kinguin got 8 world (probably fall) damage
DEBUG: tick 188715: Bomb exploded on A
```
Sample output (dust2):
```
DEBUG: tick 284853: Bomb has been planted on B by Marek77
DEBUG: tick 290102: Bomb exploded on B
DEBUG: tick 290102: B1NGO got 2 world damage (probably from the bomb explosion)
```

**Comment by markus-wa** [comment]
`DEBUG: tick 185395: (nil) hurt OLOFMEISTER * Kinguin with UNKNOWN in Head (92 hp left, 8 damage)` seems very strange. It's a couple of seconds before the explosion and a headshot.

Do you know what's up with that?

**Comment by BestAwperEver** [comment]
I also noticed this and put it in my to-look list 😄 Seems weird but might be an error on my side.

**Comment by markus-wa** [comment]
Make sure you're using v1.7.1 as well, I fixed some `EqUnknown` stuff there 🙂

**Comment by BestAwperEver** [comment]
Yep, 1.7.1 it is. Still getting the line.

**Comment by BestAwperEver** [comment]
Turns out it is just a fall damage. And that is not the only case in this particular demo (still the one in which c9 got burned by pronax's molly). I have no idea why is it counted as headshots 😄

**Comment by markus-wa** [comment]
see #178, not happy yet but let's see where it goes.

**Comment by markus-wa** [comment]
Just released `v2.0.0-beta.9` with bomb damage detection. Let me know if you notice any issues with this.

https://github.com/markus-wa/demoinfocs-golang/releases

**Comment by markus-wa** [comment]
Closing this as v2.0.0 has just been merge to master and released :partying_face:
